File: tasks/2025/round3/weights.cpp
#include <iostream>
#include <vector>
 
using namespace std;
 
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, q;
    cin >> n >> q;
    
    const int bk = 600; // bucket size
    
    vector<long long> differences(n+1), groups(1+(n+1)/bk), frq(n+1);
    differences[0] += 1LL * (n-1) * n / 2;
    groups[0] += 1LL * (n-1) * n / 2;
    
    int cntdistinct = 1;
    vector<int> distinct(bk+bk);
    distinct[1] = 1;
    frq[1] = n;
    for (int i = 1; i <= q; i++) {
        int tp;
        cin >> tp;
        
        if (tp == 1) {
            int a, b;
            cin >> a >> b;
            
            frq[a]--;
            for (int pos = 1; pos <= cntdistinct; pos++) {
                differences[abs(a - distinct[pos])] -= frq[distinct[pos]];
                groups[(abs(a - distinct[pos])) / bk] -= frq[distinct[pos]];
            }
            frq[b]--;
            for (int pos = 1; pos <= cntdistinct; pos++) {
                differences[abs(b - distinct[pos])] -= frq[distinct[pos]];
                groups[(abs(b - distinct[pos])) / bk] -= frq[distinct[pos]];
            }
            for (int pos = 1; pos <= cntdistinct; pos++) {
                differences[abs((a+b) - distinct[pos])] += frq[distinct[pos]];
                groups[(abs((a+b) - distinct[pos])) / bk] += frq[distinct[pos]];
            }
            bool found = 0;
            for (int pos = 1; pos <= cntdistinct; pos++) {
                if (distinct[pos] == a+b) {
                    found = 1;
                }
            }
            if (!found) {
                distinct[++cntdistinct] = a+b;
            }
            frq[a+b]++;
            for (int pos = cntdistinct; pos >= 1; pos--) {
                if (frq[distinct[pos]] == 0) {
                    for (int x = pos; x < cntdistinct; x++) {
                        distinct[x] = distinct[x+1];
                    }
                    cntdistinct--;
                }
            }
            
        }
        else {
            long long k;
            cin >> k;
            for (int i = 0; i <= n/bk; i++) {
                if (k - groups[i] > 0) {
                    k -= groups[i];
                }
                else {
                    for (int pos = i * bk; ; pos++) {
                        k -= differences[pos];
                        if (k <= 0) {
                            cout << pos << '\n';
                            break;
                        }
                    }
                    break;
                }
            }
        }
    }
    return 0;
}

File: tasks/2025/round3/weights.py
#!/usr/bin/env python3
# not passes even with pypy ...

from sys import stdin, stdout
input = stdin.readline
    
def solv():    
    n, q = map(int,input().split())

    bk = 600  # bucket size

    differences = [0] * (n + 1)
    groups = [0] * (1 + (n + 1) // bk)
    frq = [0] * (n + 1)

    differences[0] = groups[0] = (n - 1) * n // 2

    cntdistinct = 1
    distinct = [0] * (bk + bk)
    distinct[1] = 1
    frq[1] = n

    for _ in range(q):
        tp, *R = map(int,input().split())
        if tp == 1:
            a, b = R
            apb = a + b
            frq[a] -= 1
            for pos in range(1, cntdistinct + 1):
                abs_diff = abs(a - distinct[pos])
                differences[abs_diff] -= frq[distinct[pos]]
                groups[abs_diff // bk] -= frq[distinct[pos]]

            frq[b] -= 1
            for pos in range(1, cntdistinct + 1):
                abs_diff = abs(b - distinct[pos])
                differences[abs_diff] -= frq[distinct[pos]]
                groups[abs_diff // bk] -= frq[distinct[pos]]

            for pos in range(1, cntdistinct + 1):
                abs_diff = abs((apb) - distinct[pos])
                differences[abs_diff] += frq[distinct[pos]]
                groups[abs_diff // bk] += frq[distinct[pos]]

            found = False
            for pos in range(1, cntdistinct + 1):
                if distinct[pos] == apb:
                    found = True
                    break

            if not found:
                cntdistinct += 1
                distinct[cntdistinct] = apb

            frq[apb] += 1

            for pos in range(cntdistinct, 0, -1):
                if frq[distinct[pos]] == 0:
                    for x in range(pos, cntdistinct):
                        distinct[x] = distinct[x + 1]
                    cntdistinct -= 1

        else:
            k = R[0]
            for i in range(n // bk + 1):
                if k - groups[i] > 0:
                    k -= groups[i]
                else:
                    for pos in range(i * bk, len(differences)):
                        k -= differences[pos]
                        if k <= 0:
                            #print(pos)
                            stdout.write(str(pos)+'\n') 
                            break
                    break

solv()
